File: backend/app/nlu/parser.py
from pydantic import BaseModel, Field
from typing import Optional, List
import os
from google import genai
from google.genai import types

class ExtractedPeak(BaseModel):
    name: str = Field(description="Description or name of the peak")
    frequency_hz: float = Field(description="Frequency of the peak in Hz")

class ExtractedMachineParams(BaseModel):
    motor_rpm: Optional[float] = Field(default=None, description="Running speed in RPM")
    male_lobes: Optional[int] = Field(default=None, description="Number of lobes on male rotor")
    female_lobes: Optional[int] = Field(default=None, description="Number of lobes on female rotor")
    foundation_type: Optional[str] = Field(default=None, description="One of: 'Rigid', 'Flexible'")
    machine_group: Optional[str] = Field(default=None, description="One of: 'Group 1 (Process Gas)', 'Group 2 (Oil-Free Air)', 'Group 3 (Oil-Flooded)'")
    bearing_numbers: Optional[List[str]] = Field(default=None, description="List of bearing designations, e.g. ['7309 DE', 'NU 309']")
    measured_peaks: Optional[List[ExtractedPeak]] = Field(default=None, description="List of measured peaks with their frequencies in Hz")
    overall_vibration_rms: Optional[float] = Field(default=None, description="Overall vibration in mm/s RMS")
    
    missing_fields: List[str] = Field(default_factory=list, description="List of fields that are required but missing from the input")
    conversational_reply: Optional[str] = Field(default=None, description="Use this field to reply naturally if the user just says hello, or to ask for clarification on missing parameters.")

def parse_operator_input(text: str, expected_type: Optional[str] = None) -> ExtractedMachineParams:
    """
    Parses free-text input from the operator to extract structured parameters.
    """
    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        # Fallback for dev without API key
        return ExtractedMachineParams(
            missing_fields=["GEMINI_API_KEY_MISSING"]
        )
        
    client = genai.Client(api_key=api_key)
    
    prompt = f"""
=== PERSONA & ROLE ===
You are VDX · Diagnostic Engine, an expert NLU parser for rotating machinery and screw compressor vibration diagnostics.
Your job is to analyze the operator's input text (which may span single or multiple conversation turns) and extract telemetry parameters into the structured schema.

=== SMART INTAKE RULE ===
Check if the operator input contains ALL required diagnostic telemetry fields:
1. Speed in RPM (`motor_rpm`)
2. Male rotor lobes count (`male_lobes`)
3. Female rotor lobes count (`female_lobes`)
4. Foundation type: Rigid or Flexible (`foundation_type`)
5. Compressor Machine Group / Type: Oil-Free, Oil-Flooded, or Process Gas (`machine_group`)
6. Bearing designation / numbers (`bearing_numbers`, e.g. 7309 DE, NU 309)
7. Overall vibration in mm/s RMS (`overall_vibration_rms`)
8. Measured peak frequencies in Hz (`measured_peaks`)

- If ALL required fields are present: extract them cleanly and leave `missing_fields` empty and `conversational_reply` null.
- If ANY required fields are missing and NO telemetry data is provided (e.g. user just says "Hello" or asks a general question): populate `conversational_reply` with a natural greeting asking specifically for the missing items.
- If SOME fields are present but SOME are missing: extract all provided fields into the schema and list ONLY the unprovided items in `missing_fields`.

=== EXTRACTION RULES & HINTS ===
- Machine Group Mapping:
  - "Oil-Free" -> "Group 2 (Oil-Free Air)"
  - "Oil-Flooded" -> "Group 3 (Oil-Flooded)"
  - "Process Gas" -> "Group 1 (Process Gas)"
- Speed: Look for values followed by "RPM" or "speed".
- Lobes: Distinguish male vs female lobe counts (e.g., "4 male lobes, 6 female lobes" -> male_lobes: 4, female_lobes: 6).
- Bearings: Extract bearing model designations such as "7309 DE", "NU 309 NDE", etc.
- Peaks: Extract all frequency numbers given in Hz into `measured_peaks`.
- Strict Rule: DO NOT guess or infer values not mentioned in the input.

Operator Input History:
"{text}"
"""
    
    response = client.models.generate_content(
        model='gemini-2.5-flash',
        contents=prompt,
        config=types.GenerateContentConfig(
            response_mime_type="application/json",
            response_schema=ExtractedMachineParams,
            temperature=0.0
        ),
    )
    
    return ExtractedMachineParams.model_validate_json(response.text)

File: src/lib/nlu/parser.ts
import { GoogleGenAI, Type } from "@google/genai";
import { ExtractedMachineParams } from "../engine/models";

export async function parseOperatorInput(
  text: string,
  _expectedType?: string
): Promise<ExtractedMachineParams> {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    return {
      missing_fields: ["GEMINI_API_KEY_MISSING"],
    };
  }

  const ai = new GoogleGenAI({ apiKey });

  const prompt = `
=== PERSONA & ROLE ===
You are VDX · Diagnostic Engine, an expert NLU parser for rotating machinery and centrifugal fan vibration diagnostics.
Your job is to analyze the operator's input text (which may span single or multiple conversation turns) and extract telemetry parameters into the structured schema.

=== SMART INTAKE RULE ===
Check if the operator input contains ALL required diagnostic telemetry fields:
1. Speed in RPM (\`fan_rpm\` or \`motor_rpm\`)
2. Impeller vanes count (\`vanes\`)
3. Foundation type: Rigid or Flexible (\`foundation_type\`)
4. Machine Group / Classification according to ISO 20816-3: Group 1 (Large Machines > 300 kW) or Group 2 (Medium Machines 15-300 kW) (\`machine_group\`)
5. Bearing designation / numbers (\`bearing_numbers\`, e.g. Spherical roller bearings, 22216, 6310)
6. Overall vibration in mm/s RMS (\`overall_vibration_rms\`)
7. Measured peak frequencies in Hz (\`measured_peaks\`)

- If ALL required fields are present: extract them cleanly and leave \`missing_fields\` empty and \`conversational_reply\` null.
- If ANY required fields are missing and NO telemetry data is provided (e.g. user just says "Hello" or asks a general question): populate \`conversational_reply\` with a natural greeting asking specifically for the missing items.
- If SOME fields are present but SOME are missing: extract all provided fields into the schema and list ONLY the unprovided items in \`missing_fields\`.

=== EXTRACTION RULES & HINTS ===
- Machine Group Mapping:
  - "Group 1" or "Large" or "> 300 kW" -> "Group 1 (Large Machines > 300 kW)"
  - "Group 2" or "Medium" or "15-300 kW" -> "Group 2 (Medium Machines 15-300 kW)"
- Speed: Look for values followed by "RPM" or "speed" (e.g., 2980 RPM -> fan_rpm: 2980).
- Vanes: Look for "vanes" or "impeller vanes" (e.g., "11 vanes" -> vanes: 11).
- Bearings: Extract bearing model designations such as "Spherical roller bearings", "22216", "6310", etc.
- Foundation: Identify "Rigid" or "Flexible".
- Peaks: Extract all frequency numbers given in Hz or CPM (convert CPM to Hz if explicitly in CPM: Hz = CPM / 60) into \`measured_peaks\`.
- Strict Rule: DO NOT guess or infer values not mentioned in the input.

Operator Input History:
"${text}"
`;

  try {
    const response = await ai.models.generateContent({
      model: "gemini-2.5-flash",
      contents: prompt,
      config: {
        responseMimeType: "application/json",
        responseSchema: {
          type: Type.OBJECT,
          properties: {
            fan_rpm: { type: Type.NUMBER, description: "Fan running speed in RPM", nullable: true },
            motor_rpm: { type: Type.NUMBER, description: "Motor/Fan running speed in RPM", nullable: true },
            vanes: { type: Type.INTEGER, description: "Number of impeller vanes", nullable: true },
            foundation_type: { type: Type.STRING, description: "One of: 'Rigid', 'Flexible'", nullable: true },
            machine_group: {
              type: Type.STRING,
              description: "One of: 'Group 1 (Large Machines > 300 kW)', 'Group 2 (Medium Machines 15-300 kW)'",
              nullable: true,
            },
            bearing_numbers: {
              type: Type.ARRAY,
              items: { type: Type.STRING },
              description: "List of bearing designations, e.g. ['Spherical roller bearings', '22216']",
              nullable: true,
            },
            measured_peaks: {
              type: Type.ARRAY,
              items: {
                type: Type.OBJECT,
                properties: {
                  name: { type: Type.STRING },
                  frequency_hz: { type: Type.NUMBER },
                },
                required: ["name", "frequency_hz"],
              },
              description: "List of measured peaks with their frequencies in Hz",
              nullable: true,
            },
            overall_vibration_rms: { type: Type.NUMBER, description: "Overall vibration in mm/s RMS", nullable: true },
            missing_fields: {
              type: Type.ARRAY,
              items: { type: Type.STRING },
              description: "List of required fields missing from input",
            },
            conversational_reply: {
              type: Type.STRING,
              description: "Natural reply or clarification response if telemetry is incomplete or conversational",
              nullable: true,
            },
          },
          required: ["missing_fields"],
        },
        temperature: 0.0,
      },
    });

    const parsed: ExtractedMachineParams = JSON.parse(response.text || "{}");
    if (!parsed.missing_fields) {
      parsed.missing_fields = [];
    }
    // Normalize fan_rpm and motor_rpm
    if (parsed.fan_rpm && !parsed.motor_rpm) {
      parsed.motor_rpm = parsed.fan_rpm;
    } else if (parsed.motor_rpm && !parsed.fan_rpm) {
      parsed.fan_rpm = parsed.motor_rpm;
    }
    return parsed;
  } catch (error) {
    console.error("Gemini parse error:", error);
    return {
      missing_fields: ["PARSER_ERROR"],
      conversational_reply: `An error occurred while parsing input: ${error instanceof Error ? error.message : String(error)}`,
    };
  }
}
